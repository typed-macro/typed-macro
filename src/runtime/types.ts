import { CallExpression } from '@babel/types'
import { NodePath } from '@babel/traverse'
import { PathHelper } from '@/helper/path'
import { StateHelper } from '@/helper/state'
import { TransformHelper } from '@/helper/transform'
import { Babel } from '@/helper/babel'

export type MacroContext = {
  /**
   * The file path of the module currently being handled.
   */
  filepath: string
  /**
   * The original source code of the module currently being handled.
   */
  code: string
  /**
   * The arguments nodes of the call-macro expression currently being handled.
   */
  args: CallExpression['arguments']
  /*
   * The NodePath of the call-macro expression currently being handled.
   */
  path: NodePath
  /*
   * Whether in ssr mode.
   */
  ssr: boolean
}

export type MacroHelper = TransformHelper & PathHelper & StateHelper

export type MacroHandler = (
  /**
   * Context about the macro caller.
   */
  ctx: MacroContext,
  /**
   * Babel tools to help operate node paths and nodes.
   */
  babel: Readonly<Babel>,
  /**
   * Wrappers upon babel tools to simplify operations.
   */
  helper: Readonly<MacroHelper>
) => void

export type Macro = {
  name: string
  apply: MacroHandler
}

export type NamespacedMacros = { [namespace: string]: Macro[] }

export type NamespacedModules = { [namespace: string]: string }

export type MacroWithType = Macro & {
  __types: string
}

export function isMacroWithType(o: unknown): o is MacroWithType {
  return !!(o as MacroWithType).__types
}

export type Exportable = ({ macros: Macro[] } | { code: string }) & {
  customTypes?: string
}

export type NamespacedExportable = { [namespace: string]: Exportable }

export type NormalizedExports = {
  macros: NamespacedMacros
  modules: NamespacedModules
  types: NamespacedTypes
}

export function normalizeExports(
  exports: NamespacedExportable
): NormalizedExports {
  const macros: NamespacedMacros = Object.create(null)
  const modules: NamespacedModules = Object.create(null)
  const types: NamespacedTypes = Object.create(null)
  Object.keys(exports).forEach((ns) => {
    const item = exports[ns]
    types[ns] = {
      moduleScope: item.customTypes || '',
      macroScope: [],
    }
    if ('code' in item) {
      modules[ns] = item.code
    } else {
      macros[ns] = item.macros.map((m) => ({ name: m.name, apply: m.apply }))
      types[ns].macroScope = item.macros.map((m) => {
        if (!isMacroWithType(m)) throw new Error(`${m.name} is not a macro`)
        return m.__types
      })
    }
  })
  return {
    macros,
    modules,
    types,
  }
}

export type NamespacedTypes = {
  [namespace: string]: {
    moduleScope: string
    macroScope: string[]
  }
}

export function renderTypes(types: NamespacedTypes) {
  const namespaces = Object.keys(types)
  return namespaces
    .map((ns) => {
      const item = types[ns]
      return `declare module '${ns}' {
${[item.moduleScope, item.macroScope.join('\n')].filter((t) => !!t).join('\n')}
}`
    })
    .join('\n')
}
