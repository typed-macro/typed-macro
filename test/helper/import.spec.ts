import { getStatement } from '../testutils'
import {
  ImportOption,
  matchImportStmt,
  renderImportStmt,
} from '@/helper/import'
import { ImportDeclaration } from '@babel/types'

describe('tools#import', () => {
  it('matchImportStmt() should work with loose=true', () => {
    const stmt = getStatement(
      `import { a as _a } from 'a'`
    ) as ImportDeclaration
    const testCases: { option: ImportOption; match: boolean }[] = [
      { option: { moduleName: 'a' }, match: true } /* loose */,
      { option: { moduleName: 'b' }, match: false },
      { option: { moduleName: 'a', exportName: 'a' }, match: true } /* loose */,
      {
        option: { moduleName: 'a', exportName: 'a', localName: 'a' },
        match: false,
      },
      {
        option: { moduleName: 'a', exportName: 'a', localName: '_a' },
        match: true,
      },
      { option: { moduleName: 'a', defaultName: 'a' }, match: false },
      { option: { moduleName: 'a', namespaceName: 'a' }, match: false },
    ]
    testCases.forEach((c) =>
      expect(matchImportStmt(c.option, stmt, true /* loose */)).toBe(c.match)
    )
  })

  it('matchImportStmt() should work with loose=false', () => {
    {
      const stmt = getStatement(
        `import { a as _a } from 'a'`
      ) as ImportDeclaration
      const testCases: { option: ImportOption; match: boolean }[] = [
        { option: { moduleName: 'a' }, match: false },
        { option: { moduleName: 'b' }, match: false },
        { option: { moduleName: 'a', exportName: 'a' }, match: false },
        {
          option: { moduleName: 'a', exportName: 'a', localName: 'a' },
          match: false,
        },
        {
          option: { moduleName: 'a', exportName: 'a', localName: '_a' },
          match: true,
        },
        { option: { moduleName: 'a', defaultName: 'a' }, match: false },
        { option: { moduleName: 'a', namespaceName: 'a' }, match: false },
      ]
      testCases.forEach((c) =>
        expect(matchImportStmt(c.option, stmt, false /* loose */)).toBe(c.match)
      )
    }
    {
      const stmt = getStatement(`import { a } from 'a'`) as ImportDeclaration
      const testCases: { option: ImportOption; match: boolean }[] = [
        { option: { moduleName: 'a', exportName: 'a' }, match: true },
        {
          option: {
            moduleName: 'a',
            exportName: 'a',
            localName: 'a',
          } /* it's equal to { moduleName: 'a', exportName: 'a' } */,
          match: true,
        },
        {
          option: { moduleName: 'a', exportName: 'a', localName: '_a' },
          match: false,
        },
        { option: { moduleName: 'a', defaultName: 'a' }, match: false },
        { option: { moduleName: 'a', namespaceName: 'a' }, match: false },
      ]
      testCases.forEach((c) =>
        expect(matchImportStmt(c.option, stmt, false /* loose */)).toBe(c.match)
      )
    }
    {
      const stmt = getStatement(`import * as a from 'a'`) as ImportDeclaration
      expect(
        matchImportStmt(
          { moduleName: 'a', namespaceName: 'a' },
          stmt,
          false /* loose */
        )
      ).toBe(true)
      expect(
        matchImportStmt(
          { moduleName: 'a', defaultName: 'a' },
          stmt,
          false /* loose */
        )
      ).toBe(false)
    }
    {
      const stmt = getStatement(`import a from 'a'`) as ImportDeclaration
      expect(
        matchImportStmt(
          { moduleName: 'a', defaultName: 'a' },
          stmt,
          false /* loose */
        )
      ).toBe(true)
      expect(
        matchImportStmt(
          { moduleName: 'a', namespaceName: 'a' },
          stmt,
          false /* loose */
        )
      ).toBe(false)
    }
  })

  it('renderImportStmt() should render code properly', function () {
    const testCases: ImportOption[] = [
      { moduleName: 'a' },
      { moduleName: 'a', exportName: 'a' },
      { moduleName: 'a', exportName: 'a', localName: '_a' },
      { moduleName: 'a', defaultName: 'a' },
      { moduleName: 'a', namespaceName: 'a' },
    ]
    testCases.forEach((c) => expect(renderImportStmt(c)).toMatchSnapshot())
  })
})
