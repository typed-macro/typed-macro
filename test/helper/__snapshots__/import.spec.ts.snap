// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderImportStmt() should render code properly 1`] = `"import 'a'"`;

exports[`renderImportStmt() should render code properly 2`] = `"import { a } from 'a'"`;

exports[`renderImportStmt() should render code properly 3`] = `"import { a as _a } from 'a'"`;

exports[`renderImportStmt() should render code properly 4`] = `"import a from 'a'"`;

exports[`renderImportStmt() should render code properly 5`] = `"import * as a from 'a'"`;
